lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -928,12 +971,16 @@ internal class EvaluatingCompiler(
                                 selectProjectionThunk(projectEnv, joinedValues)
                             }
 
-                            val quantifiedRows = when(setQuantifier) {
+                            var quantifiedRows = when(setQuantifier) {
                                 // wrap the ExprValue to use ExprValue.equals as the equality
                                 SetQuantifier.DISTINCT -> projectedRows.filter(createUniqueExprValueFilter())
                                 SetQuantifier.ALL -> projectedRows
                             }
 
+                            if (limit != null) {
+                                quantifiedRows = quantifiedRows.take(compileLimit(limit, env))
```

Comment:
Since this is inside a call to `thunkFactory.thunkEnv(metas)` (line 968), the call to `compileLimit` is being invoked at *evaluation-time*, which means the limit clause is being re-compiled for every row.  To fix this move the `compileLimit` call outside of the closure, i.e. to 964.

Reply:
This is an important aspect of how `EvaluatingCompiler` works: a significant portion of the the code in it  executes at evaluation-time and not-compile time because it resides in closures (thunks) such as this one.  It is important to avoid re-compiling parts of expressions within the body of a thunk.

Reply:
I separated out the compilation ([L955](https://github.com/partiql/partiql-lang-kotlin/blob/ad2aa3e73f549c3ff66605d254c2c5b7dfdbc277/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L955)) from the evaluation ([L892](https://github.com/partiql/partiql-lang-kotlin/blob/ad2aa3e73f549c3ff66605d254c2c5b7dfdbc277/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L892)) so the compilation isn't done for every row.